File: backend-express/src/api/routes/places.ts
import { Request, Response, Router } from "express";

import { extractFindQuery, validateBody, validateQuery } from "@/lib/express_";
import { zod } from "@/lib/zod_";
import { crudsPlace } from "@/models/cruds";
import {
    PlaceGetSchema,
    PlacePost,
    PlacePostSchema,
    PlacePut,
    PlacePutSchema,
    PlaceReadSchema,
    PlaceSearchQuery,
    PlaceSearchSchema,
    PlacesPaginatedSchema,
} from "@/models/schemas";

import { swaggerRegistery } from "../docs";
import { Authenticated, getCurrentUser } from "../middlewares";

export const placeRouter = Router();

// Get Places Endpoint
async function getPlaces(req: Request, resp: Response) {
    // All places are public
    const query = req.parsedBody as PlaceSearchQuery;
    resp.status(200).json(await crudsPlace.paginate(query));
}

placeRouter.get(
    "/",
    Authenticated,
    extractFindQuery(PlaceSearchSchema, "query"),
    getPlaces
);

swaggerRegistery.registerPath({
    method: "get",
    path: "/places/",
    request: {
        query: PlaceSearchSchema,
    },
    responses: {
        200: {
            description: "Search and Filter places",
            content: {
                "application/json": {
                    schema: PlacesPaginatedSchema,
                },
            },
        },
    },
    tags: ["Place"],
    summary: "Search and Retrieve places",
    security: [{ OAuth2Password: [] }],
});

// Post search
async function searchPlaces(req: Request, resp: Response) {
    // All places are public
    const query = req.parsedBody as PlaceSearchQuery;
    resp.status(200).json(await crudsPlace.paginate(query));
}

placeRouter.post(
    "/search",
    Authenticated,
    extractFindQuery(PlaceSearchSchema, "body"),
    searchPlaces
);

swaggerRegistery.registerPath({
    method: "post",
    path: "/places/search",
    request: {
        body: {
            content: {
                "application/json": {
                    schema: PlaceSearchSchema,
                },
            },
            description: "Place advanced search",
            required: true,
        },
    },
    responses: {
        200: {
            description: "Search and Filter places",
            content: {
                "application/json": {
                    schema: PlacesPaginatedSchema,
                },
            },
        },
    },
    tags: ["Place"],
    summary: "Search and Retrieve places",
    security: [{ OAuth2Password: [] }],
});

// Post New Places
async function createPlace(req: Request, resp: Response) {
    // Use safeCreate to avoid a user posting a place for another user
    const parsed = req.parsedBody as PlacePost;
    const currentUser = getCurrentUser(req);
    const newPlace = await crudsPlace.userPost(currentUser, parsed, {
        process: true,
    });
    resp.status(200).json(newPlace);
}

placeRouter.post(
    "/",
    Authenticated,
    validateBody(PlacePostSchema),
    createPlace
);

swaggerRegistery.registerPath({
    method: "post",
    path: "/places/",
    request: {
        body: {
            content: {
                "multipart/form-data": {
                    schema: PlacePostSchema,
                },
            },
            description: "Place creation",
            required: true,
        },
    },
    responses: {
        200: {
            description: "Place creation",
            content: {
                "application/json": {
                    schema: PlaceReadSchema,
                },
            },
        },
    },
    tags: ["Place"],
    summary: "Place Creation",
    security: [{ OAuth2Password: [] }],
});

// Get a place by ID
async function getPlace(req: Request, res: Response) {
    // All places are public
    const place = await crudsPlace.get(req.params.placeId, {
        fields: req.parsedQuery.fields,
        process: true,
    });
    res.status(200).json(place);
}

placeRouter.get(
    "/:placeId",
    Authenticated,
    validateQuery(PlaceGetSchema),
    getPlace
);

swaggerRegistery.registerPath({
    method: "get",
    path: "/places/{placeId}",
    request: {
        params: zod.object({
            placeId: zod.number().openapi({
                example: 123456789,
                description: "Place Id",
            }),
        }),
        query: PlaceGetSchema,
    },
    responses: {
        200: {
            description: "Place data",
            content: {
                "application/json": {
                    schema: PlaceReadSchema,
                },
            },
        },
    },
    tags: ["Place"],
    summary: "Search and Retrieve place by id",
    security: [{ OAuth2Password: [] }],
});

// Edit Places
async function editPlace(req: Request, res: Response) {
    const parsed = req.parsedBody as PlacePut;
    const currentUser = getCurrentUser(req);
    const updatedPlace = await crudsPlace.userPut(
        currentUser,
        req.params.placeId,
        parsed,
        { process: true }
    );
    res.status(200).json(updatedPlace);
}

placeRouter.put(
    "/:placeId",
    Authenticated,
    validateBody(PlacePutSchema),
    editPlace
);

swaggerRegistery.registerPath({
    method: "put",
    path: "/places/{placeId}",
    request: {
        params: zod.object({
            placeId: zod.number().openapi({
                example: 123456789,
                description: "Place id",
            }),
        }),
        body: {
            content: {
                "application/json": {
                    schema: PlacePutSchema,
                },
            },
            description: "Place updated data",
            required: true,
        },
    },
    responses: {
        200: {
            description: "Place data",
            content: {
                "application/json": {
                    schema: PlaceReadSchema,
                },
            },
        },
    },
    tags: ["Place"],
    summary: "Update places",
    security: [{ OAuth2Password: [] }],
});

// Delete Places
async function deletePlace(req: Request, res: Response) {
    const currentUser = getCurrentUser(req);
    await crudsPlace.userDelete(currentUser, req.params.placeId);
    res.status(200).json({
        message: `Deleted place ${req.params.placeId}`,
    });
}

placeRouter.delete("/:placeId", Authenticated, deletePlace);

swaggerRegistery.registerPath({
    method: "delete",
    path: "/places/{placeId}",
    request: {
        params: zod.object({
            placeId: zod.number().openapi({
                example: 123456789,
                description: "Place Id",
            }),
        }),
    },
    responses: {
        200: {
            description: "Deletion confirmation message",
            content: {
                "application/json": {
                    schema: zod.object({
                        message: zod.string().openapi({
                            example: "Deleted place 507f1f77bcf86cd799439011",
                        }),
                    }),
                },
            },
        },
    },
    tags: ["Place"],
    summary: "Delete place by id",
    security: [{ OAuth2Password: [] }],
});

File: backend-fastapi/api/routes/places.py
from typing import Annotated, cast

from fastapi import APIRouter, Depends, Path, Query

from api.middlewares import get_current_user
from models.cruds import CrudsPlace, PlaceOptions
from models.schemas import (
    PlaceMultipartPost,
    PlacePutSchema,
    PlaceReadSchema,
    PlaceSearchSchema,
    PlaceSelectableFields,
    PlacesPaginatedSchema,
    UserReadSchema,
)

from ..middlewares import get_cruds_place

place_router = APIRouter(prefix="/api/places", tags=["Place"])

place_id_param = Path(
    ...,
    examples=["507f1f77bcf86cd799439011"],
    description="The ID of the place",
)


@place_router.get(
    "/",
    summary="Search and Filter places",
    response_model=PlacesPaginatedSchema,
)
async def get_places(
    query: Annotated[PlaceSearchSchema, Query()],
    cruds: CrudsPlace = Depends(get_cruds_place),
    _: UserReadSchema = Depends(get_current_user),
):
    options = PlaceOptions(process=True, fields=None)
    return await cruds.paginate(query.to_search(), options)


@place_router.post(
    "/search",
    summary="Search and Retrieve places",
    response_model=PlacesPaginatedSchema,
)
async def get_places_from_post(
    query: PlaceSearchSchema,
    cruds: CrudsPlace = Depends(get_cruds_place),
    _: UserReadSchema = Depends(get_current_user),
):
    return await cruds.paginate(query.to_search())


@place_router.post(
    "/", summary="Place creation", response_model=PlaceReadSchema
)
async def create_place(
    multipart_form: PlaceMultipartPost = Depends(),
    cruds: CrudsPlace = Depends(get_cruds_place),
    user: UserReadSchema = Depends(get_current_user),
):
    options = PlaceOptions(fields=None, process=True)
    return await cruds.user_post(user, multipart_form.to_post_schema(), options)


@place_router.get(
    "/{place_id}",
    summary="Search and Retrieve place by id",
    response_model=PlaceReadSchema | dict,
)
async def get_place(
    fields: Annotated[
        list[PlaceSelectableFields] | None,
        Query(
            description="Fields to include in the response; omit for full document",
            examples=[["id"]],
        ),
    ] = None,
    place_id: str = place_id_param,
    cruds: CrudsPlace = Depends(get_cruds_place),
    _: UserReadSchema = Depends(get_current_user),
):
    options = PlaceOptions(fields=fields, process=True)
    return await cruds.get_partial(place_id, options)


@place_router.put(
    "/{place_id}", summary="Update places", response_model=PlaceReadSchema
)
async def update_place(
    form: PlacePutSchema,
    place_id: str = place_id_param,
    cruds: CrudsPlace = Depends(get_cruds_place),
    current_user: UserReadSchema = Depends(get_current_user),
):
    options = PlaceOptions(fields=None, process=True)
    return await cruds.user_put(current_user, place_id, form, options)


@place_router.delete(
    "/{place_id}",
    summary="Delete place by id",
    responses={
        200: {
            "description": "Deletion confirmation message",
            "content": {
                "application/json": {
                    "example": {
                        "message": "Deleted place 507f1f77bcf86cd799439011"
                    }
                }
            },
        }
    },
)
async def delete_place(
    user: UserReadSchema = Depends(get_current_user),
    place_id: str = place_id_param,
    cruds: CrudsPlace = Depends(get_cruds_place),
):
    await cruds.user_delete(user, place_id)
    return dict(message=f"Deleted place {place_id}")
